src/main/java/net/neoforged/neoforge/capabilities/ItemCapability.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * An {@code ItemCapability} gives flexible access to objects of type {@code T} from item stacks.
+ *
+ * <h3>Querying an item capability</h3>
+ * <p>
+ * To get an object of type {@code T}, use {@link ItemStack#getCapability(ItemCapability)}.
+ * For example, to query an item handler from an item stack:
+ * 
+ * <pre>{@code
+ * ItemStack stack = ...;
+ *
+ * IItemHandler maybeHandler = stack.getCapability(ForgeCapabilities.ItemHandler.ITEM);
+ * if (maybeHandler != null) {
+ *     // Use maybeHandler
+ * }
+ * }</pre>
+ *
+ * <h3>Providing an item capability</h3>
+ * <p>
+ * To provide objects of type {@code T}, register providers to {@link RegisterCapabilitiesEvent}. For example:
+ * 
+ * <pre>{@code
+ * modBus.addListener((RegisterCapabilitiesEvent event) -> {
+ *     event.registerItem(
+ *         ForgeCapabilities.ItemHandler.ITEM, // capability to register for
+ *         (itemStack, context) -> <return the IItemHandler for the itemStack>,
+ *         MY_ITEM);
+ * });
+ * }</pre>
+ *
+ * @param <T> Type of queried objects.
+ * @param <C> Type of the additional context.
+ */
+public final class ItemCapability<T, C> {
+    public static <T, C> ItemCapability<T, C> create(ResourceLocation name, Class<T> typeClass, Class<C> contextClass) {
+        return (ItemCapability<T, C>) registry.create(name, typeClass, contextClass);
+    }
+
+    public static <T> ItemCapability<T, Void> createVoid(ResourceLocation name, Class<T> typeClass) {
+        return create(name, typeClass, void.class);
+    }
+
+    public ResourceLocation name() {
+        return name;
+    }
+
+    public Class<T> typeClass() {
+        return typeClass;
+    }
+
+    public Class<C> contextClass() {
+        return contextClass;
+    }
+
+    // INTERNAL
+    private static final CapabilityRegistry<ItemCapability<?, ?>> registry = new CapabilityRegistry<>(ItemCapability::new);
+
+    private final ResourceLocation name;
+    private final Class<T> typeClass;
+    private final Class<C> contextClass;
+
+    private ItemCapability(ResourceLocation name, Class<T> typeClass, Class<C> contextClass) {
+        this.name = name;
+        this.typeClass = typeClass;
+        this.contextClass = contextClass;
+    }
+
+    final Map<Item, List<ICapabilityProvider<ItemStack, C, T>>> providers = new IdentityHashMap<>();
```

Comment:
This is an interesting problem to solve, and I haven't found a satisfactory solution; it would also be nice for entities where e.g. curios might want to contribute to the overall player inventory.

For Fabric, I used a completely ad-hoc implementation to combine multiple handlers for fluid-containing items. This works, but it is not the cleanest:
https://github.com/FabricMC/fabric/blob/ed814d36ba32d89a54d2dac4e8f67037fb368668/fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidStorage.java#L103

Reply:
Added as future work. Marking as resolved for now.